nx/lib/nx.ex

```diff
@@ -10375,6 +10375,198 @@ defmodule Nx do
     impl!(tensor).fft(out, tensor, opts)
   end
 
+  @doc """
+  Returns a tensor with rank N+1 in which the last N dimensions
+  are sliding windows of rank N over the input tensor of rank N.
+
+  ## Options
+
+    * `:strides` - a list of length N which represents the strides
+      over each dimension. Can also be a number which will be the
+      stride over each dimension. Defaults to `1`.
+    * `:padding` - A valid padding as per `Nx.Shape.pad/2` over the
+      input tensor's shape. Defaults to `:valid`
+
+  ## Examples
+
+      iex> Nx.as_windowed(Nx.tensor([[0, 1, 2, 3, 4], [10, 11, 12, 11, 10]]), {2, 4})
+      #Nx.Tensor<
+        s64[2][2][4]
+        [
+          [
+            [0, 1, 2, 3],
+            [10, 11, 12, 11]
+          ],
+          [
+            [1, 2, 3, 4],
+            [11, 12, 11, 10]
+          ]
+        ]
+      >
+
+      iex> Nx.as_windowed(Nx.tensor([[0, 1, 2, 3, 4], [10, 11, 12, 11, 10]]), {2, 3})
+      #Nx.Tensor<
+        s64[3][2][3]
+        [
+          [
+            [0, 1, 2],
+            [10, 11, 12]
+          ],
+          [
+            [1, 2, 3],
+            [11, 12, 11]
+          ],
+          [
+            [2, 3, 4],
+            [12, 11, 10]
+          ]
+        ]
+      >
+
+      iex> t = Nx.tensor([
+      ...>  [0, 1, 2],
+      ...>  [3, 4, 5],
+      ...>  [-1, -2, -3],
+      ...>  [-4, -5, -6],
+      ...>  [-7, -8, -9]
+      ...> ])
+      iex> Nx.as_windowed(t, {3, 2}, strides: [2, 1])
+      #Nx.Tensor<
+        s64[4][3][2]
+        [
+          [
+            [0, 1],
+            [3, 4],
+            [-1, -2]
+          ],
+          [
+            [1, 2],
+            [4, 5],
+            [-2, -3]
+          ],
+          [
+            [-1, -2],
+            [-4, -5],
+            [-7, -8]
+          ],
+          [
+            [-2, -3],
+            [-5, -6],
+            [-8, -9]
+          ]
+        ]
+      >
+  """
+  def as_windowed(tensor, window_dimensions, opts \\ []) do
```

Comment:
@josevalim Are we keeping this here even though it can be implemented through pure Nx?
I'm almost certain the `for` can be turned into a `while` for a better `defn` version which could be used
as the `optional` default implementation.

For binary backend and torchx we can do better than this through binary parsing and through Torchx.unfold.
However, I'm also happy with keeping this in NxSignal instead.